landlab/graph/framed_voronoi/framed_voronoi.py

```diff
@@ -0,0 +1,483 @@
+        return np.concatenate(HorizontalRectVoronoiGraph.nodes_at_edge(shape))
+
+    @staticmethod
+    def nodes_at_edge(shape):
+        """
+        Parameters
+        ----------
+        shape: tuple of int
+            Number of rows and number of columns
+
+        Returns
+        -------
+        tuple(ndarray of Int)
+            For each edge give the ids of the nodes present at the edge
+
+        Examples
+        --------
+        >>> from landlab.graph.framed_voronoi.framed_voronoi import HorizontalRectVoronoiGraph
+        >>> HorizontalRectVoronoiGraph.nodes_at_edge((3, 3))
+        (array([2, 5]), array([8, 7]), array([6, 3]), array([0, 1]))
+        """
+        n_rows, n_cols = shape
+        if n_rows == n_cols == 1:
+            return (np.array([0]),) + (np.array([], dtype=int),) * 3
+        (
+            northeast,
+            northwest,
+            southwest,
+            southeast,
+        ) = HorizontalRectVoronoiGraph.corner_nodes(shape)
+
+        if n_rows > 1:
+            south = np.arange(southwest, southeast)
+        else:
+            south = np.array([southwest], dtype=int)
+
+        if n_cols > 1:
+            west = np.arange(northwest, southwest, -n_cols)
+        else:
+            west = np.array([northwest], dtype=int)
+
+        return (
+            np.arange(southeast, northeast, n_cols),
+            np.arange(northeast, northwest, -1),
+            west,
+            south,
+        )
+
+
+class FramedVoronoiGraph(DelaunayGraph):
+    """Graph of a unstructured grid of Voronoi Delaunay cells and
+    irregular patches. It is a special type of VoronoiDelaunay graph in which
+    the initial set of points is arranged in a fixed lattice (e.g. like a rectangular
+    raster grid) named here "layout" and the core points are then moved aroung their
+    initial position by a random distance, lower than a certain threshold.
+
+    Examples
+    --------
+    >>> from landlab.graph import FramedVoronoiGraph
+
+    >>> graph = FramedVoronoiGraph((3, 3), node_layout='rect', random_seed=False)
+    >>> graph.number_of_nodes
+    9
+
+    >>> graph.x_of_node[2:5]    # doctest: +NORMALIZE_WHITESPACE
+    array([ 2.        ,  0.        ,  1.07341707])
+    >>> graph.y_of_node[2:6]    # doctest: +NORMALIZE_WHITESPACE
+    array([ 0.        ,  0.749     ,  1.03337157,  1.251     ])
+
+    >>> graph.number_of_links
+    16
+    >>> graph.number_of_patches
+    8
+    """
+
+    def __init__(
+        self,
+        shape,
+        xy_spacing=(1.0, 1.0),
+        xy_of_lower_left=(0.0, 0.0),
+        orientation="horizontal",
+        node_layout="rect",
+        sort=False,
+        xy_min_spacing=(0.5, 0.5),
+        random_seed=True,
+        seed=(200, 500),
+    ):
+        """Create the graph.
+
+        Parameters
+        ----------
+        shape : int or tuple of int
+            For a rectangular layout, number of rows and columns of nodes.
+            If int, rows number = columns number = value
```

Comment:
I don't think we use this pattern for any other grid (i.e. that *shape* can be an int or a tuple), do we? I think shape is always a *tuple*.

Reply:
now requires tuple of int only